usrsctplib/netinet/sctp_callout.c

```diff
@@ -75,72 +76,201 @@ uint32_t sctp_get_tick_count(void) {
 
 /*
  * SCTP_TIMERQ_LOCK protects:
- * - SCTP_BASE_INFO(callqueue)
- * - sctp_os_timer_next: next timer to check
+ * - SCTP_BASE_INFO(timers_queue)
  * - sctp_os_timer_current: current callout callback in progress
  * - sctp_os_timer_current_tid: current callout thread id in progress
- * - sctp_os_timer_waiting: some thread is waiting for callout to complete
- * - sctp_os_timer_wait_ctr: incremented every time a thread wants to wait
- *                           for a callout to complete.
+ * - sctp_os_timer_current_changed: conditional variable signaled when
+ *                                  current callout pointer is changed
  */
-static sctp_os_timer_t *sctp_os_timer_next = NULL;
```

Comment:
Note for reviewers:
It is no longer necessary to store pointer to next timer as global variable and keep it in sync when timer's list state changed (in `sctp_os_timer_start`/`sctp_os_timer_stop`/`sctp_handle_tick`). This is because linked binary heap is now used to store timers and it maintains state to provide next soonest timer to iterate over all expired timers.

It is still necessary to store never de-referenced pointer to current timer in `sctp_os_timer_current`. The pointer to current timer is used to determine currently running timer. This "executing" flag can not be stored in timer's state, because timer can self-detroy in callback and there is no way to reset "executing" flag in timer's memory, because it was already freed.